distributed/metrics.py

```diff
@@ -103,3 +108,170 @@ def resync(self) -> None:
+        The callback must accept the same parameters as :meth:`digest_metric`.
+        """
+        cbs = self._callbacks.get()
+        tok = self._callbacks.set(cbs + [callback])
+        try:
+            yield
+        finally:
+            tok.var.reset(tok)
+
+    def digest_metric(self, label: str, value: float, unit: str) -> None:
+        """Invoke the currently set context callbacks for an arbitrary quantitative
+        metric.
+        """
+        cbs = self._callbacks.get()
+        for cb in cbs:
+            cb(label, value, unit)
+
+    def meter(
+        self,
+        label: str,
+        unit: str = "seconds",
+        func: Callable[[], float] = timemod.perf_counter,
+        floor: float | Literal[False] = 0.0,
+    ) -> Any:
+        """Convenience context manager or function decorator which calls func() before
+        and after the wrapped code, calculates the delta, and finally calls
+        :meth:`digest_metric`. It also subtracts any other calls to :meth:`meter` or
+        :meth:`digest_metric` with the same unit performed within the context, so that
+        the total is strictly additive.
+
+        Parameters
+        ----------
+        label: str
+            label to pass to the callback
+        unit: str, optional
+            unit to pass to the callback. Default: seconds
+        func: callable
+            see :func:`meter`
+        floor: bool, optional
+            see :func:`meter`
+        """
+        parent = self
+
+        class _ContextMeter_meter:
+            def __call__(self, f2):
+                @wraps(f2)
+                def wrapper(*args, **kwargs):
+                    with self:
+                        return f2(*args, **kwargs)
+
+                return wrapper
+
+            def _callback(self, label2: str, value2: float, unit2: str) -> None:
+                if unit2 == unit:
+                    # This must be threadsafe to support when callbacks are invoked from
+                    # distributed.utils.offload; '+=' on a float would not be
+                    # threadsafe!
+                    self.offsets.append(value2)
+
+            def __enter__(self):
+                self.offsets = []
+                self.cb_ctx = parent.add_callback(self._callback)
+                self.cb_ctx.__enter__()
+                self.meter_ctx = meter(func, floor=False)
+                self.meter_output = self.meter_ctx.__enter__()
+
+            def __exit__(self, exc_type, exc_val, exc_tb):
+                self.cb_ctx.__exit__(exc_type, exc_val, exc_tb)
+                self.meter_ctx.__exit__(exc_type, exc_val, exc_tb)
+                delta = self.meter_output.delta - sum(self.offsets)
+                if floor is not False:
+                    delta = max(floor, delta)
+                parent.digest_metric(label, delta, unit)
+                del self.offsets
+                del self.cb_ctx
+                del self.meter_ctx
+
+        return _ContextMeter_meter()
```

Comment:
Instead of this ugly-looking local class I could have written a general-purpose `@contextmanager_or_decorator`  decorator, but it felt like overkill - particularly after looking at the internals of the implementation of `contextlib.contextmanager`.